libavfilter/x86/vf_gblur.asm

```diff
@@ -20,17 +20,371 @@
 
 %include "libavutil/x86/x86util.asm"
 
+SECTION .data
+
+gblur_vindex_width: dd 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15
+gblur_transpose_16x16_mask: dw 0xcc, 0x33, 0xaa, 0x55, 0xaaaa, 0x5555
+gblur_transpose_16x16_indices1: dq 2, 3, 0, 1, 6, 7, 4, 5
+gblur_transpose_16x16_indices2: dq 1, 0, 3, 2, 5, 4, 7, 6
+gblur_transpose_16x16_indices3: dd 1, 0, 3, 2, 5, 4, 7, 6, 9, 8, 11, 10, 13, 12, 15, 14
+
 SECTION .text
 
 ; void ff_horiz_slice_sse4(float *ptr, int width, int height, int steps,
 ;                          float nu, float bscale)
 
 %macro HORIZ_SLICE 0
-%if UNIX64
+%if cpuflag(avx512)
+cglobal horiz_slice_asm, 5, 12, 32, buffer, width, height, steps, localbuf, x, y, step, stride, remain, ptr
+%elif UNIX64
 cglobal horiz_slice, 4, 9, 9, ptr, width, height, steps, x, y, step, stride, remain
 %else
 cglobal horiz_slice, 4, 9, 9, ptr, width, height, steps, nu, bscale, x, y, step, stride, remain
 %endif
+%if cpuflag(avx512)
+    VBROADCASTSS zm0, xmm0 ; nu
+    VBROADCASTSS zm1, xmm1 ; bscale
+    vpbroadcastd zm2, widthd
+    vpmulld zm2, zm2, [gblur_vindex_width] ; vindex width
+
+    kmovw   k6, [gblur_transpose_16x16_mask + 0*2]
+    kmovw   k5, [gblur_transpose_16x16_mask + 1*2]
+    kmovw   k4, [gblur_transpose_16x16_mask + 2*2]
+    kmovw   k3, [gblur_transpose_16x16_mask + 3*2]
+    kmovw   k2, [gblur_transpose_16x16_mask + 4*2]
+    kmovw   k1, [gblur_transpose_16x16_mask + 5*2]
+    movu  zm28, [gblur_transpose_16x16_indices1]
+    movu  zm29, [gblur_transpose_16x16_indices2]
+    movu  zm30, [gblur_transpose_16x16_indices3]
+
+    xor yq, yq ; y = 0
+    xor xq, xq ; x = 0
+
+    cmp heightq, 0x10
+    jl .y_scalar
+    sub heightq, 0x10
+    .loop_y:
+        xor stepd, stepd
+        .loop_step:
+            mov  ptrq, yq
+            imul ptrq, widthq
+            lea  ptrq, [bufferq + ptrq*4] ; ptr = buffer + y * width;
+            
+            kxnorw  k7, k7, k7
+            vgatherdps zm3{k7}, [ptrq + zm2*4]
+            mulps zm3, zm1
+            movu [localbufq], zm3
+            inc xq
+            add ptrq, 4 ; ptr++
+
+            ; Filter rightwards
+            .loop_x:
+                kxnorw k7, k7, k7
+                vgatherdps zm4{k7}, [ptrq + zm2*4]
+                FMULADD_PS zm3, zm3, zm0, zm4, zm3
+                add localbufq, 0x40
+                movu [localbufq], zm3
+                add ptrq, 4
+
+                inc xq
+                cmp xq, widthq
+                jl .loop_x
+
+            sub ptrq, 4
+            mulps zm3, zm1
+            kxnorw k7, k7, k7
+            vscatterdps [ptrq + zm2*4]{k7}, zm3
+
+            ; Filter leftwards
+            dec xq
+            .loop_x_back:
+                ; Read 16x16 matrix from local buffer
```

Comment:
not sure we can use some trick like this  or not
https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=transpose&expand=3186,5961,5961